packages/core/src/app/pg-core/db/chain-db.ts

```diff
@@ -1,24 +1,523 @@
+        await this.removeMerkleTreeNode(blockNumber, siblingIndex)
+      }
+    }
+  }
+
+  /**
+   * Creates an inclusion proof for a given transaction.
+   * @param transaction Transaction to create a proof for.
+   * @returns the inclusion proof for that transaction.
+   */
+  public async getInclusionProof(
+    transaction: Transaction
+  ): Promise<InclusionProof> {
+    const blockNumber = transaction.block
+    const leafIndex = await this.getLeafIndex(transaction)
+    const treeHeight = await this.getBlockTreeHeight(blockNumber)
+
+    // Get the list of siblings up the tree.
+    const siblingIndices: number[] = getMerkleSiblingIndices(
+      leafIndex,
+      treeHeight
+    )
+
+    // Get the nodes necessary to generate the proof.
+    const proof: string[] = []
+    for (const siblingIndex of siblingIndices) {
+      const sibling = await this.getMerkleTreeNode(blockNumber, siblingIndex)
+
+      // Don't have the sibling, can't compute the proof.
+      if (sibling === null) {
+        throw new Error('Cannot compute inclusion proof, missing sibling node.')
+      }
+
+      proof.push(sibling)
+    }
+
+    return proof
+  }
+
+  /**
+   * Adds a block hash to the database.
+   * @param blockNumber Block to add a hash for.
+   * @param blockHash Hash to add.
+   */
+  public async addBlockHash(
+    blockNumber: number,
+    blockHash: string
+  ): Promise<void> {
+    const key = KEYS.BLOCKS.encode([blockNumber])
+    const value = Buffer.from(blockHash)
+    await this.db.put(key, value)
+  }
+
+  /**
+   * Inserts the tree height for a given block.
+   * @param blockNumber Block number to add tree height for.
+   * @param treeHeight Height of the tree.
+   */
+  public async addBlockTreeHeight(
+    blockNumber: number,
+    treeHeight: number
+  ): Promise<void> {
+    const key = KEYS.TREE_HEIGHTS.encode([blockNumber])
+    const value = Buffer.allocUnsafe(4)
+    value.writeUInt32BE(treeHeight, 0)
+    await this.db.put(key, value)
+  }
+
+  /**
+   * Queries the tree height for a given block.
+   * @param blockNumber Block number to query.
+   * @returns the tree height for that block.
+   */
+  public async getBlockTreeHeight(blockNumber: number): Promise<number> {
+    const key = KEYS.TREE_HEIGHTS.encode([blockNumber])
+    const value = await this.db.get(key)
+    return value.readUInt32BE(0)
+  }
+
+  /**
+   * Adds a transaction to the database.
+   * @param transaction Transaction to add.
+   */
+  public async addTransaction(transaction: Transaction): Promise<void> {
+    const end = getTransactionRangeEnd(transaction)
+    const key = KEYS.TRANSACTIONS.encode([transaction.block, end])
+    const value = encode(transaction)
+    await this.db.put(key, value)
+  }
+
+  /**
+   * Sets the leaf index for a given transaction.
+   * @param transaction Transaction to set.
+   * @param leafIndex Leaf index for the transaction.
+   */
+  public async addLeafIndex(
```

Comment:
Need to store the leaf indices of each transaction. @karlfloersch is this OK or do you think there's a cleaner way to store this?

Reply:
Ah yes I believe you have to store the index of the transaction. There might be a cleaner way but I can't think of it off the top of my head.

One weird note is that I actually was using the start position as the key, seeking to the end of the range I care about, and then doing a reverse iteration. Tbh I'm not sure why I did that instead of storing it by the end, seeking with the start, and using a forward iteration. However, just note that's what I did and there *may* be a reason but I'm not sure.